src/libraries/System.Private.CoreLib/src/System/Delegate.cs

```diff
@@ -164,27 +164,25 @@ public bool MoveNext()
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static bool operator ==(Delegate? d1, Delegate? d2)
         {
-            // Test d2 first to allow branch elimination when inlined for null checks (== null)
-            // so it can become a simple test
-            if (d2 is null)
+            if (ReferenceEquals(d1, d2))
             {
-                return d1 is null;
+                return true;
             }
 
-            return ReferenceEquals(d2, d1) ? true : d2.Equals((object?)d1);
+            return d2 is not null && d2.Equals(d1);
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static bool operator !=(Delegate? d1, Delegate? d2)
         {
             // Test d2 first to allow branch elimination when inlined for not null checks (!= null)
             // so it can become a simple test
```

Comment:
These comments aren't really "correct" anymore

We should likely ensure that the same general codegen and dead code elimination still happens when inlined (for RyuJIT and Mono).

Reply:
> We should likely ensure that the same general codegen and dead code elimination still happens 

+1. The MihuBot report has number of regressions. The regressions seem to be caused by dead code elimination no longer happening - a lot of new `System.Object:Equals` calls that did not exist before.


!!!

Reply:
The list of regressions seems to be a good data point about methods that use `== null` / `!= null`, but that should better be using `is null` / `is not null` instead.

Reply:
> The list of regressions seems to be a good data point about methods that use == null / != null, but that should better be using is null / is not null instead.

Are you interested in a submitting PR that fixes these? (Just in the shipping libraries, test can and should stay as is.)

Reply:
> use `== null` / `!= null`, but that should better be using `is null` / `is not null` instead.

Are custom `.Equals` implementation that might return `true` with a single `null` (like for example with Unity) not a concern here?

Reply:
We can look at it case-by-case. Most of the affected operator implementations check for the null upfront and won't call `Equals` with null.

Do you have an example of where it may be a problem?